File: typescript/timestep/utils/io.ts
/** I/O utilities. */

import { writeFileSync, mkdirSync } from 'fs';
import { dirname } from 'path';

export function now(): number {
  return Date.now() / 1000;
}

export function clamp01(x: number): number {
  return Math.max(0.0, Math.min(1.0, x));
}

export function writeJson(path: string, obj: any): void {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, JSON.stringify(obj, null, 2), 'utf-8');
}

File: python/timestep/utils/io.py
"""I/O utilities."""

import json
import time
from pathlib import Path
from typing import Any


def now() -> float:
    """Get current timestamp."""
    return time.time()


def clamp01(x: float) -> float:
    """Clamp value to [0, 1] range."""
    return max(0.0, min(1.0, x))


def write_json(path: Path, obj: Any) -> None:
    """Write JSON object to file."""
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(obj, indent=2, ensure_ascii=False), encoding="utf-8")
